File: Android/ProjectColor/app/src/main/java/com/example/projectcolor/components/Checksum.kt
package com.example.projectcolor.components

/**
 * onesComplement is a function that computes the one's complement of a given binary string. The one's complement is obtained by inverting each bit in the string.
 *
 * **Parameters:**
 *
 * - `data`: A `String` representing the binary data for which the one's complement will be calculated.
 *
 * **Returns:**
 *
 * - `String`: Returns a new binary string where each bit is the inverse of the corresponding bit in the input string. Specifically, '0' becomes '1' and '1' becomes '0'.
 *
 * **Functionality:**
 *
 * - The function iterates over each character in the input string.
 * - For each character, it inverts the bit: '0' is changed to '1', and '1' is changed to '0'.
 * - The inverted bits are stored in a `CharArray`, which is then converted back into a string and returned.
 *
 * - This function is typically used as part of a checksum calculation to help verify data integrity by ensuring that the transmitted data can be checked for errors.
 */
fun onesComplement(data: String): String {
    val result = CharArray(data.length)
    for (index in data.indices) {
        result[index] = if (data[index] == '0') '1' else '0'
    }
    return result.concatToString()
}

/**
 * binaryAdd is a function that performs binary addition of two binary strings, each representing a block of bits, and returns the result.
 * The function also handles any carry-over that results from the addition, ensuring that the final binary string is accurate.
 *
 * **Parameters:**
 *
 * - `result`: A `String` representing the first binary block to be added. This is typically the cumulative result of previous additions.
 * - `nextBlock`: A `String` representing the next binary block to be added to the `result`.
 * - `blockSize`: An `Int` specifying the size of the blocks (i.e., the number of bits) involved in the binary addition.
 *
 * **Returns:**
 *
 * - `String`: Returns the binary string that results from the addition of `result` and `nextBlock`, taking into account any carry-over.
 *
 * **Functionality:**
 *
 * - The function initializes an array of characters (`resultArray`) from the `result` string, which will hold the result of the addition.
 * - Starting from the least significant bit (rightmost), it iterates through each bit in the `result` and `nextBlock`, performs binary addition, and accounts for carry.
 * - If there is any carry remaining after the initial pass, the function continues to propagate the carry through the more significant bits (moving left).
 * - The final result, which includes any carry-over, is returned as a new binary string.
 *
 * - This function is typically used in the context of checksum calculations, where multiple binary blocks are summed to create a final checksum value.
 */
fun binaryAdd(result: String, nextBlock: String, blockSize: Int): String {
    val resultArray = result.toCharArray()
    var carry = 0
    for (k in blockSize - 1 downTo 0) {
        val sum = (nextBlock[k] - '0') + (resultArray[k] - '0') + carry
        resultArray[k] = (sum % 2 + '0'.toInt()).toChar()
        carry = sum / 2
    }

    // Handle the carry
    if (carry != 0) {
        for (l in blockSize - 1 downTo 0) {
            val sum = (resultArray[l] - '0') + carry
            resultArray[l] = (sum % 2 + '0'.toInt()).toChar()
            carry = sum / 2
            if (carry == 0) break
        }
    }
    return resultArray.concatToString()
}

/**
 * checkSum is a function that calculates the checksum of a given binary string by breaking it into blocks, performing binary addition on each block,
 * and then applying a one's complement to the result. This checksum is typically used to verify the integrity of data during transmission.
 *
 * **Parameters:**
 *
 * - `data`: A `String` representing the binary data for which the checksum will be calculated.
 * - `blockSize`: An `Int` representing the size of the blocks in which the binary data will be divided for the checksum calculation.
 *
 * **Returns:**
 *
 * - `String`: Returns the one's complement of the binary addition result as a binary string, which serves as the checksum.
 *
 * **Functionality:**
 *
 * - The function pads the input binary string with leading zeros if its length is not a multiple of `blockSize`.
 * - It initializes the result with the first block of the padded data.
 * - The function then iterates over the remaining blocks, performing binary addition of each block to the cumulative result.
 * - After processing all blocks, the function calculates the one's complement of the result, which is returned as the checksum.
 *
 * - This function is typically used in data integrity checks, where the checksum is appended to data before transmission, allowing the receiver to verify that the data has not been altered.
 */
fun checkSum(data: String, blockSize: Int): String {
    var paddedData = data
    val n = data.length

    // Pad with leading zeros if necessary
    val padSize = blockSize - (n % blockSize)
    if (padSize != blockSize) {
        paddedData = "0".repeat(padSize) + data
    }

    // Initialize the result with the first block
    var result = paddedData.substring(0, blockSize)

    // Process each block
    for (i in blockSize until paddedData.length step blockSize) {
        result = binaryAdd(result, paddedData.substring(i, i + blockSize), blockSize)
    }

    // Compute one's complement
    return onesComplement(result)
}

/**
 * hexStringToBinaryString is a function that converts a hexadecimal string into its equivalent binary string representation.
 * Each hexadecimal digit is translated into a 4-bit binary sequence.
 *
 * **Parameters:**
 *
 * - `hexString`: A `String` representing the hexadecimal data to be converted to binary.
 *
 * **Returns:**
 *
 * - `String`: Returns a binary string equivalent to the input hexadecimal string, where each hex digit is converted into a 4-bit binary sequence.
 *
 * **Functionality:**
 *
 * - The function iterates over each character in the input `hexString`.
 * - For each character, it appends the corresponding 4-bit binary string to the `binaryString` variable.
 * - The conversion accounts for both uppercase and lowercase hexadecimal digits.
 * - If an invalid hexadecimal character is encountered, the function throws an `IllegalArgumentException`.
 *
 * - This function is typically used when binary data needs to be processed in its binary form but is originally provided in a hexadecimal format.
 */
fun hexStringToBinaryString(hexString: String): String {
    var binaryString = ""
    for (char in hexString) {
        binaryString += when (char) {
            '0' -> "0000"
            '1' -> "0001"
            '2' -> "0010"
            '3' -> "0011"
            '4' -> "0100"
            '5' -> "0101"
            '6' -> "0110"
            '7' -> "0111"
            '8' -> "1000"
            '9' -> "1001"
            'a', 'A' -> "1010"
            'b', 'B' -> "1011"
            'c', 'C' -> "1100"
            'd', 'D' -> "1101"
            'e', 'E' -> "1110"
            'f', 'F' -> "1111"
            else -> throw IllegalArgumentException("Invalid hex character")
        }
    }
    return binaryString
}

File: Arduino/tests/checksum.cpp
#include <stdio.h>
#include <bits/stdc++.h>
#include <iostream>
#include <string>
using namespace std;

// Function to find the One's complement
// of the given binary string
string Ones_complement(string data)
{
	for (int i = 0; i < data.length(); i++) {
		if (data[i] == '0')
			data[i] = '1';
		else
			data[i] = '0';
	}

	return data;
}

// Function to return the checksum value of
// the given string when divided in K size blocks
string checkSum(string data, int block_size)
{
	// Check data size is divisible by block_size
	// Otherwise add '0' front of the data
	int n = data.length();
	if (n % block_size != 0) {
		int pad_size = block_size - (n % block_size);
		for (int i = 0; i < pad_size; i++) {
			data = '0' + data;
		}
	}

	// Binary addition of all blocks with carry
	string result = "";

	// First block of data stored in result variable
	for (int i = 0; i < block_size; i++) {
		result += data[i];
	}

	// Loop to calculate the block
	// wise addition of data
	for (int i = block_size; i < n; i += block_size) {

		// Stores the data of the next block
		string next_block = "";

		for (int j = i; j < i + block_size; j++) {
			next_block += data[j];
		}

		// Stores the binary addition of two blocks
		string additions = "";
		int sum = 0, carry = 0;

		// Loop to calculate the binary addition of
		// the current two blocks of k size
		for (int k = block_size - 1; k >= 0; k--) {
			sum += (next_block[k] - '0')
			       + (result[k] - '0');
			carry = sum / 2;
			if (sum == 0) {
				additions = '0' + additions;
				sum = carry;
			}
			else if (sum == 1) {
				additions = '1' + additions;
				sum = carry;
			}
			else if (sum == 2) {
				additions = '0' + additions;
				sum = carry;
			}
			else {
				additions = '1' + additions;
				sum = carry;
			}
		}

		// After binary add of two blocks with carry,
		// if carry is 1 then apply binary addition
		string final = "";

		if (carry == 1) {
			for (int l = additions.length() - 1; l >= 0;
			        l--) {
				if (carry == 0) {
					final = additions[l] + final;
				}
				else if (((additions[l] - '0') + carry) % 2
				         == 0) {
					final = "0" + final;
					carry = 1;
				}
				else {
					final = "1" + final;
					carry = 0;
				}
			}

			result = final;
		}
		else {
			result = additions;
		}
	}

	// Return One's complements of result value
	// which represents the required checksum value
	return Ones_complement(result);
}

// Function to check if the received message
// is same as the senders message
bool checker(string sent_message,
             string rec_message,
             int block_size)
{

	// Checksum Value of the senders message
	string sender_checksum
	    = checkSum(sent_message, block_size);

	// Checksum value for the receivers message
	string receiver_checksum 
	    = checkSum(rec_message + sender_checksum, block_size);

	// If receivers checksum value is 0
	if (count(receiver_checksum.begin(),
	          receiver_checksum.end(), '0')
	        == block_size) {
		return true;
	}
	else {
		return false;
	}
}


string hexstring_to_binarystring(string hexstring) {
    string binaryString = "";
    for(int index = 0; index < hexstring.length(); index ++) {
        switch(hexstring[index]) {
            case '0':
                binaryString += "0000";
                break;
            case '1':
                binaryString += "0001";
                break;
            case '2':
                binaryString += "0010";
                break;
            case '3':
                binaryString += "0011";
                break;
            case '4':
                binaryString += "0100";
                break;
            case '5':
                binaryString += "0101";
                break;
            case '6':
                binaryString += "0110";
                break;
            case '7':
                binaryString += "0111";
                break;
            case '8':
                binaryString += "1000";
                break;
            case '9':
                binaryString += "1001";
                break;
            case 'a':
                binaryString += "1010";
                break;
            case 'b':
                binaryString += "1011";
                break;
            case 'c':
                binaryString += "1100";
                break;
            case 'd':
                binaryString += "1101";
                break;
            case 'e':
                binaryString += "1110";
                break;
            case 'f':
                binaryString += "1111";
                break;
        }
    }
    return binaryString;
}

///////////////////////////////////////////////////
//TESTS

//the source file
void TEST_checksum_demo() {
	string sent_message
	    = "10000101011000111001010011101101";
	string recv_message
	    = "10000101011000111001010011101101";
	int block_size = 8;

	if (checker(sent_message,
	            recv_message,
	            block_size)) {
		cout << "No Error";
	}
	else {
		cout << "Error";
	}
}

void checksum_sandbox() {
	string sent_message
	    = "10100101101001011010010110100101";
	int block_size = 32;
	string sent_message_checksum = checkSum(sent_message, block_size);
	cout << "sent_message: " << sent_message << "\n";
	cout << "block_size: " << block_size << "\n";
    cout << "sent_message_checksum: " << sent_message_checksum << "\n";
    
    string receiver_checksum = checkSum(sent_message + sent_message_checksum, block_size);
    cout << "receiver_checksum: " << receiver_checksum << "\n";
}

void test_hexstring_conversion() {
    string hex = "f0f06e6e";
    string binaryString = hexstring_to_binarystring(hex);
    cout << "hex: " << hex << "\n";
    cout << "converted to: " << hexstring_to_binarystring(hex) << "\n";
}

void test_hex_pixel_checksum() {
    int block_size = 8;
    string sent_pixelData = "01ff66ff";
    string sent_checksum = "98";
    string sent_pixelData_binary = hexstring_to_binarystring(sent_pixelData);
    string sent_checksum_binary = hexstring_to_binarystring(sent_checksum);

    string local_pixelData = "01ff66ff";
    string local_pixelData_binary = hexstring_to_binarystring(local_pixelData);
    string local_checksum = checkSum(local_pixelData_binary, block_size);
    
    string output = checkSum(local_pixelData_binary + sent_checksum_binary, block_size);
    
    cout << "sent_pixelData(hex):\t\t" << sent_pixelData << "\n";
    cout << "sent_pixelData(binary):\t\t" << sent_pixelData_binary << "\n";
    cout << "sent_checksum:\t\t" << sent_checksum_binary << "\n";
    
    cout << "local_pixelData(hex):\t\t" << local_pixelData << "\n";
    cout << "local_pixelData(binary):\t" << local_pixelData_binary << "\n";
    cout << "local_checksum:\t\t" << local_checksum << "\n";
    
    cout << "output:\t\t\t" << output << "\n";
}

///////////////////////////////////////////////////
int main()
{
	test_hex_pixel_checksum();
	return 0;
}
